src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java

```diff
@@ -0,0 +1,412 @@
+        HttpExecutionResult result = executeHttpRequest(timeout);
+        Optional<Integer> statuscode = Optional.empty();
+        if (result.response().isPresent())
+            statuscode = Optional.of(result.response().get().statusCode());
+
+        if (result.successful() && this.config.parseResults()) { // 2xx
+            // process result
+            if (!responseBodyProcessor.add(result.actualContentLength().getAsLong(), result.hash().getAsLong(), result.outputStream().get())) {
+                this.responseBodybbaos = result.outputStream().get();
+            } else {
+                this.responseBodybbaos = new BigByteArrayOutputStream();
+            }
+        }
+
+        try {
+            this.responseBodybbaos.reset();
+        } catch (IOException e) {
+            this.responseBodybbaos = new BigByteArrayOutputStream();
+        }
+
+        if (!result.completed() && discardOnFailure) {
+            LOGGER.debug("{}\t:: Discarded execution, because the time limit has been reached: [queryID={}]", this, result.queryID);
+            return null;
+        }
+
+        return new ExecutionStats(
+                result.queryID(),
+                result.requestStart(),
+                result.duration(),
+                statuscode,
+                result.actualContentLength(),
+                result.hash,
+                result.exception()
+        );
+    }
+
+
+    private HttpExecutionResult executeHttpRequest(Duration timeout) {
+        final QueryHandler.QueryStreamWrapper queryHandle;
+        try {
+            queryHandle = config().queries().getNextQueryStream(this.querySelector);
+        } catch (IOException e) {
+            return new HttpExecutionResult(
+                this.querySelector.getCurrentIndex(),
+                Optional.empty(),
+                Instant.now(),
+                Duration.ZERO,
+                Optional.empty(),
+                OptionalLong.empty(),
+                OptionalLong.empty(),
+                Optional.of(e)
+            );
+        }
+
+        final HttpRequest request;
+
+        try {
+            request = requestFactory.buildHttpRequest(
+                    queryHandle.queryInputStream(),
+                    timeout,
+                    config().connection(),
+                    config().acceptHeader()
+            );
+        } catch (IOException | URISyntaxException e) {
+            return new HttpExecutionResult(
+                    queryHandle.index(),
+                    Optional.empty(),
+                    Instant.now(),
+                    Duration.ZERO,
+                    Optional.empty(),
+                    OptionalLong.empty(),
+                    OptionalLong.empty(),
+                    Optional.of(e)
+            );
+        }
+
+        if (((ThreadPoolExecutor) this.httpClient.executor().get()).getActiveCount() != 0) {
+            ((ThreadPoolExecutor) this.httpClient.executor().get()).shutdownNow();
```

Comment:
not tested this but I would guess that if you shutdown the executor you need to create a new one. Also, it doesn't guarantee to shutdown the running tasks: "There are no guarantees beyond best-effort attempts to stop processing actively executing tasks.". so, it would be better to wait and write the waiting time to the log. Log should contain: thread id, time when waiting started, current time. Sad but true: we can't catch tasks gone rouge.

Reply:
Won't there be a new executor created with the buildHttpClient() method?

Reply:
Also, the shutdownNow() method is supposed to call interrupt() on all the threads it has. And if I'm not mistaken the http response input stream is supposed to respond to the interrupt flag (it looks like there was acutally a [bug ](https://bugs.openjdk.org/browse/JDK-8294047) related to this). So maybe this should actually never block indefinetely?

Reply:
The bug report mentions that the bug is closed only in openjdk 20. We should document that somewhere.